drivers/storage/portworx/deployment.go

```diff
@@ -510,26 +573,32 @@ func (t *template) getArguments() []string {
 		}
 
 	} else if t.cluster.Spec.CloudStorage != nil {
-		if t.cluster.Spec.CloudStorage.DeviceSpecs != nil {
+		if t.cloudConfig != nil && len(t.cloudConfig.CloudStorage) > 0 {
+			// CapacitySpecs have higher preference over DeviceSpecs
+			for _, cloudNodeSpec := range t.cloudConfig.CloudStorage {
+				args = append(args, "-s", t.getCloudStorageArguments(cloudNodeSpec))
+			}
+		} else if t.cluster.Spec.CloudStorage.DeviceSpecs != nil {
 			for _, dev := range *t.cluster.Spec.CloudStorage.DeviceSpecs {
 				args = append(args, "-s", dev)
 			}
 		}
 		if t.cluster.Spec.CloudStorage.JournalDeviceSpec != nil &&
-			*t.cluster.Spec.CloudStorage.JournalDeviceSpec != "" {
+			len(*t.cluster.Spec.CloudStorage.JournalDeviceSpec) > 0 {
 			args = append(args, "-j", *t.cluster.Spec.CloudStorage.JournalDeviceSpec)
 		}
 		if t.cluster.Spec.CloudStorage.SystemMdDeviceSpec != nil &&
-			*t.cluster.Spec.CloudStorage.SystemMdDeviceSpec != "" {
+			len(*t.cluster.Spec.CloudStorage.SystemMdDeviceSpec) > 0 {
 			args = append(args, "-metadata", *t.cluster.Spec.CloudStorage.SystemMdDeviceSpec)
 		}
-		if t.cluster.Spec.CloudStorage.MaxStorageNodes != nil {
+		if t.cluster.Spec.CloudStorage.MaxStorageNodes != nil &&
+			*t.cluster.Spec.CloudStorage.MaxStorageNodes > 0 {
 			args = append(args, "-max_drive_set_count",
 				strconv.Itoa(int(*t.cluster.Spec.CloudStorage.MaxStorageNodes)))
 		}
-		if t.cluster.Spec.CloudStorage.MaxStorageNodesPerZone != nil {
+		if t.cloudConfig != nil && t.cloudConfig.StorageInstancesPerZone > 0 {
```

Comment:
If cloudConfig is nil, we should still pass the `cluster.Spec.CloudStorage.MaxStorageNodesPerZone` for existing consumers.